librocksdb_sys/crocksdb/c.cc

```diff
@@ -2915,4 +2927,341 @@ const char* crocksdb_pinnableslice_value(const crocksdb_pinnableslice_t* v,
+}
+
+unsigned char crocksdb_user_collected_properties_iter_valid(
+  crocksdb_user_collected_properties_iterator_t* it) {
+  return it->iter_ != it->rep_->end();
+}
+
+void crocksdb_user_collected_properties_iter_next(
+  crocksdb_user_collected_properties_iterator_t* it) {
+  (it->iter_)++;
+}
+
+const char* crocksdb_user_collected_properties_iter_key(
+  crocksdb_user_collected_properties_iterator_t* it, size_t* klen) {
+  if (klen) {
+    *klen = it->iter_->first.size();
+  }
+  return it->iter_->first.data();
+}
+
+const char* crocksdb_user_collected_properties_iter_value(
+  crocksdb_user_collected_properties_iterator_t* it, size_t* vlen) {
+  if (vlen) {
+    *vlen = it->iter_->second.size();
+  }
+  return it->iter_->second.data();
+}
+
+struct crocksdb_table_properties_t {
+  TableProperties* rep_ = nullptr;
+  crocksdb_user_collected_properties_t user_props_;
+};
+
+uint64_t crocksdb_table_properties_get_u64(crocksdb_table_properties_t* props,
+                                           crocksdb_table_property_t prop) {
+  auto rep = props->rep_;
+  switch (prop) {
+  case DATA_SIZE: return rep->data_size;
+  case INDEX_SIZE: return rep->index_size;
+  case FILTER_SIZE: return rep->filter_size;
+  case RAW_KEY_SIZE: return rep->raw_key_size;
+  case RAW_VALUE_SIZE: return rep->raw_value_size;
+  case NUM_DATA_BLOCKS: return rep->num_data_blocks;
+  case NUM_ENTRIES: return rep->num_entries;
+  case FORMAT_VERSION: return rep->format_version;
+  case FIXED_KEY_LEN: return rep->data_size;
+  case COLUMN_FAMILY_ID: return rep->column_family_id;
+  }
+  return 0;
+}
+
+const char* crocksdb_table_properties_get_str(crocksdb_table_properties_t* props,
+                                              crocksdb_table_property_t prop, size_t* slen) {
+  auto rep = props->rep_;
+  switch (prop) {
+  case COLUMN_FAMILY_NAME:
+    if (slen) *slen = rep->column_family_name.size();
+    return rep->column_family_name.data();
+  case FILTER_POLICY_NAME:
+    if (slen) *slen = rep->filter_policy_name.size();
+    return rep->filter_policy_name.data();
+  case COMPARATOR_NAME:
+    if (slen) *slen = rep->comparator_name.size();
+    return rep->comparator_name.data();
+  case MERGE_OPERATOR_NAME:
+    if (slen) *slen = rep->merge_operator_name.size();
+    return rep->merge_operator_name.data();
+  case PREFIX_EXTRACTOR_NAME:
+    if (slen) *slen = rep->prefix_extractor_name.size();
+    return rep->prefix_extractor_name.data();
+  case PROPERTY_COLLECTORS_NAMES:
+    if (slen) *slen = rep->property_collectors_names.size();
+    return rep->property_collectors_names.data();
+  case COMPRESSION_NAME:
+    if (slen) *slen = rep->compression_name.size();
+    return rep->compression_name.data();
+  }
+  return "";
+}
+
+crocksdb_user_collected_properties_t*
+crocksdb_table_properties_get_user_properties(crocksdb_table_properties_t* props) {
+  props->user_props_.rep_ =
+    const_cast<UserCollectedProperties*>(&props->rep_->user_collected_properties);
```

Comment:
Yep, `crocksdb_user_collected_properties_add` need it to be mutable, so I can't just make it const or not const.